data/src/main/java/org/apache/iceberg/data/parquet/GenericParquetWriter.java

```diff
@@ -284,6 +291,11 @@ public void write(int repetitionLevel, byte[] value) {
     }
   }
 
+  public interface StructWriterFactory {
```

Comment:
Instead of adding this interface, what about adding `WriteBuilder.createStructWriter` to do this? That way, the Flink builder could just inherit from the generic builder and override that one method.

I think that would be cleaner because adding this interface requires also adding public methods to pass the factory. I'd rather not add those public methods if we can avoid it by adding a protected method and change the builder to a protected class.

Reply:
It could be `protected`, right?

Reply:
`protected` is not enough for the outside flink package, the `WriterBuilder` is a static class in `org.apache.iceberg.data.parquet` , it must be `public` so that other packages could inherit it.

Reply:
You're right. The class should be public, but the method that will be overridden should be protected.

Reply:
Em, I've implemented two versions. the first version provides a  `public`  interfaces with protected `StructureWriterFactory` and `StructReaderFactory`, the second version use the inherit `WriterBuilder`  & `ReaderBuilder` https://github.com/apache/iceberg/pull/1125. For me , seems the first version looks much more concise. I plan to change to version#1 (with the  `public` interface and `protected` methods ). Thanks.

Reply:
Could you post the branch with the inheritance version? I'd like to see it to compare. I don't like the extra public classes and methods, and I think that would be a cleaner API. I'm curious why you think the other approach looks more concise, though.

Reply:
The inheritance version is [here](https://github.com/apache/iceberg/pull/1125/files/f4799a0000639f662c3df823240c2e27d0734b10).

> I'm curious why you think the other approach looks more concise, though.

Because the inheritance version will need to expose the `buildReader` and `buildWriter` logic , for example [this](https://github.com/apache/iceberg/pull/1125/files/f4799a0000639f662c3df823240c2e27d0734b10#diff-aa5e57d56a7b1137c0bd6b1370a1ccaaR42). If we change those logic, will also need to change in flink reader/writers. seems the same logic with the GenericParquetReader.

Reply:
Okay, how about inheriting from the outer class, then? That way, the interface and the method that accepts the factory could be `protected`. Mainly, I don't think these should be public.